src/Storages/Hive/StorageHive.cpp

```diff
@@ -163,8 +166,26 @@ class StorageHiveSource : public SourceWithProgress, WithContext
                 std::unique_ptr<ReadBuffer> raw_read_buf;
                 try
                 {
-                    raw_read_buf = std::make_unique<ReadBufferFromHDFS>(
-                        hdfs_namenode_url, current_path, getContext()->getGlobalContext()->getConfigRef());
+                    auto get_raw_read_buf = [&]() -> std::unique_ptr<ReadBuffer>
+                    {
+                        auto buf = std::make_unique<ReadBufferFromHDFS>(
+                            hdfs_namenode_url, current_path, getContext()->getGlobalContext()->getConfigRef());
+
+                        bool thread_pool_read = read_settings.remote_fs_method == RemoteFSReadMethod::threadpool;
+                        if (thread_pool_read)
+                        {
+                            return std::make_unique<AsynchronousReadBufferFromHDFS>(StorageHive::getThreadPoolReader(), read_settings, std::move(buf));
```

Comment:
Did you consider using ParallelReadBuffer? It could perform better than prefetch for large files as it splits the file into file segments and each segment is downloaded by a thread from the pool so the download is actually parallel, with prefetch it is not like this.

Reply:
I'am going to implement parallel download in another pr. Let's leave this pr for threadpool reading.